### PR TITLE
Some Psalm coverage for phpseclib3\Math\

### PR DESCRIPTION
**Pull request by hc-jworman**
Fixed some Psalm errors in phpseclib3\Math\\*.
I got to a decent stopping point and just baselined the remaining errors (see build/psalm_baseline.xml).

### CODE REVIEW

File: phpseclib/Math/BigInteger.php [hunk 1]
```diff
@@ -44,21 +45,14 @@ class BigInteger
...
+     * @var class-string<Engine>
      */
     private static $mainEngine;
 
-    /**
-     * Modular Exponentiation Engine
-     *
-     * @var string
-     */
-    private static $modexpEngine;
```
Comment (hc-jworman) [inline phpseclib/Math/BigInteger.php:56]: Only written to, but never read.

File: phpseclib/Math/BigInteger/Engines/Engine.php [hunk 2]
```diff
@@ -30,10 +30,59 @@
...
+        839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947,
+        953, 967, 971, 977, 983, 991, 997,
+    ];
+
+    /**
+     * BigInteger(0)
+     *
+     * @var array<class-string<static>, static>
+     */
+    protected static $zero = [];
```
Comment (hc-jworman) [inline phpseclib/Math/BigInteger/Engines/Engine.php:52]: Housing all of the sub-classes static variables like this, allow us to work around an error with using static::$zero (and other properties) in the parent class. The error was due to it not being possible to determine if the implementing class had the static property.

**Review by terrafrost — COMMENTED** [review]
Looks good except for the psalm_baseline.xml stuff, as I commented on in-line

File: build/psalm_baseline.xml [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+      <code>$class::randomRange($min-&gt;value, $max-&gt;value)</code>
+      <code>$class::randomRangePrime($min-&gt;value, $max-&gt;value)</code>
+      <code>$class::scan1divide($r-&gt;value)</code>
+    </UndefinedMethod>
+  </file>
+  <file src="../phpseclib/Math/BigInteger/Engines/BCMath.php">
+    <InvalidArgument occurrences="31">
+      <code>$current</code>
+      <code>$current</code>
+      <code>$current</code>
```
[the remark this answers is not in the document]
Comment (hc-jworman) [reply build/psalm_baseline.xml:19]: `$this->value` is shared by all the engines, and therefore it has a very broad and unhelpful type: `GMP|array<array-key, mixed>|int|string`.

Many of the remaining errors are due to this.

File: build/psalm_baseline.xml [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+  <file src="../phpseclib/Math/BigInteger/Engines/BCMath/BuiltIn.php">
+    <InvalidArgument occurrences="3">
+      <code>$e-&gt;value</code>
+      <code>$n-&gt;value</code>
+      <code>$x-&gt;value</code>
+    </InvalidArgument>
+  </file>
+  <file src="../phpseclib/Math/BigInteger/Engines/Engine.php">
+    <InvalidArrayAccess occurrences="3">
+      <code>$x[0]</code>
```
[the remark this answers is not in the document]
Comment (hc-jworman) [reply build/psalm_baseline.xml:77]: `$x` is `int|string`, so we would need to narrow the type to just `string` to remove this error.

File: build/psalm_baseline.xml [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+      <code>$this-&gt;value</code>
+      <code>$this-&gt;value</code>
+      <code>$y-&gt;value</code>
+    </UndefinedMethod>
+  </file>
+  <file src="../phpseclib/Math/BigInteger/Engines/BCMath/Base.php">
+    <UndefinedMethod occurrences="3">
+      <code>static::reduce($x, $n)</code>
+      <code>static::reduce(bcmul($x, $x), $n)</code>
+      <code>static::reduce(bcmul($x, $y), $n)</code>
```
[the remark this answers is not in the document]
Comment (hc-jworman) [reply build/psalm_baseline.xml:65]: #2 was my plan as well.

**Comment by hc-jworman** [comment]
The baseline was not meant to be permanent. I was just breaking up the changes into chunks to make the PRs more digestible.

**Comment by terrafrost** [comment]
I cherry picked each of the non-merge commits into the 3.0 branch and merged into master.

idk if I've ever cherry picked multiple commits before but there were also conflicts with #1751 (which is in master but not 3.0) and #1760. You probably resolved the merge conflicts with #1760 yourself but the work to resolve them would have been lost when I did the cherry picking.

**Pull request closed without merging**